core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java

```diff
@@ -13,22 +13,49 @@
  */
 package io.trino.spi.classloader;
 
-import java.io.Closeable;
-
 public class ThreadContextClassLoader
-        implements Closeable
 {
-    private final ClassLoader originalThreadContextClassLoader;
+    private ThreadContextClassLoader() {}
+
+    public static void withClassLoader(Class<?> clazz, Runnable runnable)
+    {
+        withClassLoader(clazz.getClassLoader(), runnable);
+    }
+
+    public static <T, E extends Exception> T withClassLoader(Class<?> clazz, ThrowingSupplier<T, E> supplier)
+    {
+        return withClassLoader(clazz.getClassLoader(), supplier);
+    }
+
+    public static void withClassLoader(ClassLoader newThreadContextClassLoader, Runnable runnable)
+    {
+        withClassLoader(newThreadContextClassLoader, () -> {
+            runnable.run();
+            return null;
+        });
+    }
 
-    public ThreadContextClassLoader(ClassLoader newThreadContextClassLoader)
+    public static <T, E extends Exception> T withClassLoader(ClassLoader newThreadContextClassLoader, ThrowingSupplier<T, E> supplier)
     {
-        this.originalThreadContextClassLoader = Thread.currentThread().getContextClassLoader();
+        final ClassLoader originalThreadContextClassLoader = Thread.currentThread().getContextClassLoader();
         Thread.currentThread().setContextClassLoader(newThreadContextClassLoader);
+        try {
+            return supplier.getOrThrow();
+        }
+        catch (RuntimeException e) {
+            throw e;
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+        finally {
+            Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
+        }
     }
 
-    @Override
-    public void close()
+    @FunctionalInterface
+    public interface ThrowingSupplier<T, E extends Exception>
     {
-        Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
+        T getOrThrow() throws E;
```

Comment:
If checked is used then `orThrow()` is obvious. I would skip that.

Reply:
done: 248144248448ea297f075c119f15e4e1a83eeced